[CI] Fix the build

sfgeorge opened the pull request:
- [x] Make the build green ✅ 
- [x] Test modern Rubies
- [x] Backport RSpec fixes from `develop` branch

Please forgive me this large commit in the midst:
* [SPEC] Convert specs automatically using transpec [BREAKS] e4b72e5

...For what it's worth, I didn't add anything special to that commit beyond just letting `transpec` do its thing to automatically upgrade from RSpec 2.x to 3.x syntax, so no need to review that commit with a fine-toothed comb.

benlangfeld committed:
We don't use EventMachine

Fixes #35

benlangfeld committed:
Bump copyright year

benlangfeld committed:
Typo

benlangfeld committed:
[SPEC] Use simplified assertions from commit: "Make Stream usable in a supervision tree"

See d9b806a74af498683a992f009b3882c1563cd806

benlangfeld committed:
Make sure the test suite doesn't hang

benlangfeld committed:
Avoid rspec deprecations

benlangfeld committed:
Missed some RSpec deprecations

benlangfeld committed:
The test server failing to shut down shouldn't itself fail specs

benlangfeld committed:
We don't support Celluloid 0.17 yet

sfgeorge committed:
[CI] exhibit that the build is currently broken

sfgeorge committed:
Use rake < 11.x until we upgrade RSpec past 2.x

sfgeorge committed:
Appease ruby-1.9.3 by holding cucumber back as < 3.0

sfgeorge committed:
[CI] Test more Rubies

sfgeorge committed:
[CI] Relax Bundler development dependency.

Should get things passing on:
- ruby-2.5.3
- jruby-9.1.17.0

bklang commented on .travis.yml:
```diff
@@ -1,15 +1,27 @@
     - rvm: rbx-2.1.1
+env:
+  global:
+  - JRUBY_OPTS='--debug'
```
Are you still working on this? Or do you intend to keep this flag? (and if so, why?)

sfgeorge replied:
Good q. This is not needed yet unless we have coveralls on, which we don't.  Since it doesn't add value, I'll drop this. 

Oh and yes, turns out I still have some commits to add.. coming soon!

sfgeorge commented:
@bklang @lpradovera @gfaza Finally ready for review.  Can you review this for me?

sfgeorge commented:
@bklang @lpradovera @gfaza Sorry, not ready for review yet.  My future intent is, once we've reviewed & settled on changes here, to have the `support/2.x` branch merged into `develop`.

However, my changes won't cleanly merge-up to `develop` because I didn't realize that @benlangfeld already tackled some of these kinds of changes on `develop` in commits like 2df64170561808c674b997cea61ca3aedd50e307 and 4a8f6985354c96f378e21c5a843cf6efc75c2840.

I'm going to re-work this branch to incorporate @benlangfeld's changes so that it will merge ~cleanly~ _mostly cleanly_ to develop later on with no surprises.

sfgeorge committed:
[CI] Always run cucumber in strict mode

sfgeorge committed:
[SPEC] Don't error on RSpec deprecations.

Makes it easier to ensure that the build stays green from the RSpec 2.99 -> RSpec 3.x transition.

sfgeorge committed:
Convert specs to RSpec 2.99.2 syntax with Transpec

This conversion is done by Transpec 3.4.0 with the following command:
    transpec -f

* 65 conversions
    from: obj.should
      to: expect(obj).to

* 51 conversions
    from: == expected
      to: eq(expected)

* 8 conversions
    from: it { should ... }
      to: it { is_expected.to ... }

* 6 conversions
    from: obj.should_not
      to: expect(obj).not_to

* 5 conversions
    from: =~ /pattern/
      to: match(/pattern/)

* 5 conversions
    from: it { should_not ... }
      to: it { is_expected.not_to ... }

* 3 conversions
    from: obj.should_receive(:message)
      to: expect(obj).to receive(:message)

* 2 conversions
    from: be_false
      to: be_falsey

* 2 conversions
    from: its(:attr) { }
      to: describe '#attr' do subject { super().attr }; it { } end

* 1 conversion
    from: lambda { }.should
      to: expect { }.to

* 1 conversion
    from: mock('something')
      to: double('something')

* 1 conversion
    from: obj.stub(:message)
      to: allow(obj).to receive(:message)

* 1 conversion
    from: pending
      to: skip

For more details: https://github.com/yujinakayama/transpec#supported-conversions

sfgeorge committed:
[SPEC] mark one skipped test as expected to fail

sfgeorge committed:
[SPEC] Replace undocumented, unsupported "feature" of #with

...with a documented, supported argument matcher block to #receive.

Unsupported in #with:
https://github.com/rspec/rspec-mocks/issues/377#issuecomment-21492096

Supported in #receive:
https://github.com/rspec/rspec-mocks#arbitrary-handling

sfgeorge committed:
[SPEC] Upgrade to RSpec 3.x

sfgeorge committed:
[SPEC] Run transpec one more time to fix deprecations

Convert specs to RSpec 3.8.2 syntax with Transpec

This conversion is done by Transpec 3.4.0 with the following command:
    transpec -f

* 2 conversions
    from: obj.stub(:message => value)
      to: allow(obj).to receive_messages(:message => value)

For more details: https://github.com/yujinakayama/transpec#supported-conversions

sfgeorge committed:
Revert "Use rake < 11.x until we upgrade RSpec past 2.x"

This reverts commit a9d85a51ed3917ee5551aaa7f798cde660eaa18d.

sfgeorge commented:
@bklang @lpradovera @gfaza After a few tweaks this is ready for review  😊 😬

bklang reviewed (approved):
Wow, that was a lot of work! Thank you for your efforts here to modernize the test suite. Cheers! 🥂

bklang commented:
@sfgeorge merging because this is clearly useful & helpful. But I also noticed that this is on the `support/2.x` branch. Is this useful on the main development branch as well?

lpradovera commented:
This is absolutely amazing!

sfgeorge commented:
> Wow, that was a lot of work! Thank you for your efforts here to modernize the test suite. Cheers! 🥂

> This is absolutely amazing!

Thanks guys!  You rock!

> @sfgeorge merging because this is clearly useful & helpful. But I also noticed that this is on the support/2.x branch. Is this useful on the main development branch as well?

@bklang Absolutely.  I wanted to 1st settle any potential PR changes, but you made it easy for me.  I'll now open another PR to up-port the changes to the develop branch, settling some minor/inevitable git conflicts along the way.

sfgeorge commented:
Also, thank you @benlangfeld for having made a bunch of these stabilization fixes.  That would have been A LOT harder to do if I didn't have the benefit of plucking a bunch of the fixes you already made. 🙇
